File: programmers/전력망을 둘로 나누기.py
from collections import defaultdict, deque


def bfs(start, graph, visited):
    queue = deque([start])
    visited[start] = True
    cnt = 1
    while queue:
        q = queue.popleft()
        for g in graph[q]:
            if not visited[g]:
                visited[g] = True
                queue.append(g)
                cnt += 1
    return cnt


def solution(n, wires):
    min_diff = 99

    for i in range(len(wires)):
        graph = defaultdict(list)
        for j, (v1, v2) in enumerate(wires):
            if i == j:
                continue
            graph[v1].append(v2)
            graph[v2].append(v1)

        visited = [False] * (n + 1)
        # 전체 - (1번 wire와 연결되어있는 wire개수)
        f = bfs(1, graph, visited)
        e = n - f

        min_diff = min(min_diff, abs(f - e))

    return min_diff

File: programmers/전력망을 둘로 나누기.js
function solution(n, wires) {
    let minAns = n
    for (let i=0; i<wires.length; i++) {
        const graph = Array.from({ length: n + 1 }, () => [])
        
        for (j=0; j<wires.length; j++) {
            if (j === i) continue
            
            graph[wires[j][1]].push(wires[j][0])
            graph[wires[j][0]].push(wires[j][1])
        }
        
        const bfs = (s) => {
            const queue = [s]
            let head = 0
            const visited = Array(n + 1).fill(false)
            visited[s] = true
            
            let cnt = 0
            while (head < queue.length) {
                const q = queue[head++]
                cnt++
                
                for (const nq of graph[q]) {
                    if (!visited[nq]) {
                        queue.push(nq)
                        visited[nq] = true
                    }
                }
            }
            return cnt
        }
        
        // 전체 - (1번 wire와 연결되어있는 wire개수)
        const a = bfs(1)
        const b = n - a
        minAns = Math.min(minAns, Math.abs(a - b))
    }
    
    return minAns;
}